ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/core/PendingTransactions.java

```diff
@@ -51,28 +55,71 @@
   private final Map<Address, SortedMap<Long, TransactionInfo>> transactionsBySender =
       new HashMap<>();
 
-  private final Collection<PendingTransactionListener> listeners =
-      newSetFromMap(new ConcurrentHashMap<>());
+  private final Subscribers<PendingTransactionListener> listeners = new Subscribers<>();
+
+  private final Subscribers<PendingTransactionDroppedListener> transactionDroppedListeners =
+      new Subscribers<>();
 
   private final int maxPendingTransactions;
   private final Clock clock;
+  private final Counter localTransactionAddedCounter;
 
-  public PendingTransactions(final int maxPendingTransactions, final Clock clock) {
+  private final Counter remoteTransactionAddedCounter;
+  private Map<DecrementTransactionKey, Counter> transactionRemovedCounters = new HashMap<>();
+
+  public PendingTransactions(
+      final int maxPendingTransactions, final Clock clock, final MetricsSystem metricsSystem) {
     this.maxPendingTransactions = maxPendingTransactions;
     this.clock = clock;
+    final LabelledMetric<Counter> transactionCounter =
+        metricsSystem.createLabelledCounter(
+            MetricCategory.TRANSACTION_POOL,
+            "transactions_total",
+            "Count of transactions changed in the transaction pool",
+            "source");
+    localTransactionAddedCounter = transactionCounter.labels("local", "added");
+    addRemovedCounter(transactionCounter, "local", "addedToBlock");
+    addRemovedCounter(transactionCounter, "local", "dropped");
+
+    remoteTransactionAddedCounter = transactionCounter.labels("remote", "added");
+    addRemovedCounter(transactionCounter, "remote", "addedToBlock");
+    addRemovedCounter(transactionCounter, "remote", "dropped");
+  }
+
+  private void addRemovedCounter(
+      final LabelledMetric<Counter> transactionCounter,
+      final String source,
+      final String addedToBlock) {
+    transactionRemovedCounters.put(
+        new DecrementTransactionKey("local".equals(source), "addedToBlock".equals(addedToBlock)),
+        transactionCounter.labels(source, "removed", addedToBlock));
```

Comment:
I think you've probably over-optimised this (probably because of my previous suggestions, sorry). It's ok to call `.labels` multiple times. It's very slightly faster if you can just store the result like with `localTransactionAddedCounter` and `removeTransactionAddedCounter`.  Given there's both a local/remote and addedToBlock/dropped axis, I'd just store the `LabelledMetric<Counter>` and then when you need to increment call `.labels(...).inc()`. The indirection of looking up in a map likely nullifies any benefit to the caching.